src/bls12_381/fq2.rs

```diff
@@ -55,6 +55,11 @@ impl Fq2 {
 
         t1
     }
+
+    pub fn hash(k: &[u8], nonce: &[u8]) -> Self {
+        let half = nonce.len() / 2;
+        Fq2 { c0: Fq::hash(k, &nonce[ .. half]), c1: Fq::hash(k, &nonce[half ..]) }
```

Comment:
If I read the field abstraction spaghetti correctly, 0x00/0xff  tag from sw_encode in nonce[32] always ends up in c1, and c0 in turn having same hash for both "half points" which sw_encode adds together later. This probably isn't what was intended.